### DISCUSSION TITLE
Struct copy semantics

### DISCUSSION THREAD

**Discussion by ntrel** [Neat-Lang/neat#22] [General]
Hi,
I'm trying to understand the following:
```d
module dtor;

import std.stdio;

struct S
{
    int i;
    this(S){print("copy");}
    ~this(){print("dtor");}
}

void f(S s) {"f: $(s.i)".print;}

void main()
{
    mut S s;
    {
        S t = s; // copy
        s.i++;
        "$(s.i), $(t.i)".print;
        // dtor
    }
    "call:".print;
    s.f; // doesn't call copy or dtor
    "end".print;
}
```
Why does calling `f` avoid calling the copy constructor and destructor, but `t` doesn't?

**Comment by FeepingCreature** [comment]
That's probably a bug. I'm not exactly sure which way it's a bug, but that looks like a bug at first glance.

I think because `s` is mutable, `s.f` should (has to) make a copy.

**Reply by FeepingCreature** [reply]
Yep, it's definitely a bug. Got a repro where it clobbers a visible reference. Fixing.

**Comment by FeepingCreature** [comment]
Fixed on master: https://github.com/Neat-Lang/neat/commit/7527d82c319010fa120120b6fac9605a9a295365
